Lib/email/utils.py

```diff
@@ -103,12 +104,126 @@ def formataddr(pair, charset='utf-8'):
+    start = None
+    remove = []
+    for pos, ch in _iter_escaped_chars(addr):
+        if ch == '"':
+            if start is None:
+                start = pos
+            else:
+                remove.append((start, pos))
+                start = None
+
+    result = []
+    pos = 0
+    for start, stop in remove:
+        result.append(addr[pos:start])
+        pos = stop + 1
+    result.append(addr[pos:])
+
+    return ''.join(result)
+
+
+def getaddresses(fieldvalues, *, strict=True):
+    """Return a list of (REALNAME, EMAIL) or ('','') for each fieldvalue.
+
+    When parsing fails for a fieldvalue, a 2-tuple of ('', '') is returned in
+    its place.
 
-def getaddresses(fieldvalues):
-    """Return a list of (REALNAME, EMAIL) for each fieldvalue."""
-    all = COMMASPACE.join(str(v) for v in fieldvalues)
-    a = _AddressList(all)
-    return a.addresslist
+    If strict is True, use a strict parser which rejects malformed inputs.
+    In this case, if the resulting list of parsed addresses is greater than
+    the number of fieldvalues in the input list, a parsing error has occurred
+    and consequently a list containing a single empty 2-tuple [('', '')] is
+    returned in its place. This is done to avoid invalid output.
+
+    Malformed input: getaddresses(['alice@example.com <bob@example.com>'])
+    Invalid output: [('', 'alice@example.com'), ('', 'bob@example.com')]
+    Safe output: [('', '')]
+
+    """
+    if not strict:
+        all = COMMASPACE.join(str(v) for v in fieldvalues)
+        a = _AddressList(all)
+        return a.addresslist
+
+    fieldvalues = [str(v) for v in fieldvalues]
+    fieldvalues = _pre_parse_validation(fieldvalues)
+    addr = COMMASPACE.join(v for v in fieldvalues)
+    a = _AddressList(addr)
+    result = _post_parse_validation(a.addresslist)
+
+    # Treat output as invalid if the number of addresses is not equal to the
+    # expected number of addresses.
+    n = 0
+    for v in fieldvalues:
+        # When a comma is used in the Real Name part it is not a deliminator.
+        # So strip those out before counting the commas.
+        v = _strip_quoted_realnames(v)
+        # Expected number of addresses: 1 + number of commas
+        n += 1 + v.count(',')
+    if len(result) != n:
+        return [('', '')]
+
+    return result
+
+
+def _check_parenthesis(addr):
+    # Ignore parenthesis in quoted real names.
+    addr = _strip_quoted_realnames(addr)
+
+    opens = 0
+    for pos, ch in _iter_escaped_chars(addr):
+        if ch == '(':
+            opens += 1
+        elif ch == ')':
+            opens -= 1
+            if opens < 0:
+                return False
+    return (opens == 0)
+
+
+def _pre_parse_validation(email_header_fields):
+    accepted_values = []
+    for v in email_header_fields:
+        if not _check_parenthesis(v):
+            v = "('', '')"
+        accepted_values.append(v)
+
+    return accepted_values
+
+
+def _post_parse_validation(parsed_email_header_tuples):
```

Comment:
`_pre_parse_validation` and `_post_parse_validation` can use list comprehensions or be turned into generator functions.

Reply:
_post_parse_validation() output should be a list. There is no advantage to convert it to a generator.

If _pre_parse_validation() becomes a generator, parseaddr() has to create a list anyway for `list[0]`. getaddresses() cannot use a generator for _pre_parse_validation() result neither, since it consumes fieldvalues twice.